### PR TITLE
App Submission: Bitcoin Knots

### PR DESCRIPTION
**Pull request by Retropex**
# App Submission

### App name
Bitcoin Knots

### 256x256 SVG icon

https://svgshare.com/i/12oJ.svg

### Gallery images

Delegated to Umbrel.

### I have tested my app on:
- [x] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [x] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

### CODE REVIEW

**Comment by Retropex** [comment]
Hello,

I would like to submit an alternative implementation to Bitcoin Core, but I run into complications with UX because I would like Knots to be fully compatible with applications that are dependent on Core but I am afraid that this will cause conflict.

I also have a problem with digest of docker images, in the docker file compose we can only put one digest but the repository contains several images for each architecture I don't know which one I should put.

Thank you in advance for your help.

**Comment by smolgrrr** [comment]
Thanks for the submission! Really great work here, I actually recently added your community app-store based on this to to my list https://github.com/smolgrrr/awesome-umbrel/blob/main/README.md

What you're looking for is the digest of the manifest list. I usually use this to find it:
`docker run mplatform/manifest-tool inspect retropexx/umbrel-bitcoin-knots:v0.5.0`

Will test this out now, and we'll investigate the UX problem you mention

**Comment by nmfretz** [comment]
Hey @Retropex, thanks for working on this and submitting. We've been waiting for an alternate Bitcoin Node app to make its way to the app store. Congratulations on being the first!

> I would like to submit an alternative implementation to Bitcoin Core, but I run into complications with UX because I would like Knots to be fully compatible with applications that are dependent on Core but I am afraid that this will cause conflict.

We are planning to implement "optional dependencies" for apps in a future umbrelOS update. This will mean that allow users to let other apps know which Bitcoin Node backend to connect to. 

> I also have a problem with digest of docker images, in the docker file compose we can only put one digest but the repository contains several images for each architecture I don't know which one I should put.

image

**Comment by nmfretz** [comment]
Hey @Retropex, thanks for working on this and submitting. We've been waiting for an alternate Bitcoin Node app to make its way to the app store. Congratulations on being the first!

We've got our heads down at the moment on other aspects of our codebase, but we should have time to properly review and help you avoid IP/port clashes with the other Bitcoin Node app soon. Thanks for your patience.

One thing to check for now is that all Umbrel logos are removed from your `retropexx/umbrel-bitcoin-knots:v0.5.0` UI fork and that we differentiate it from Umbrel's Bitcoin Node app in the app description. This is so that your app is properly differentiated from Umbrel's Bitcoin Node app and so that users are aware that Umbrel is not the official maintainer of the app. I believe we only have the logo in the Loading component: https://github.com/getumbrel/umbrel-bitcoin/blob/master/ui/src/components/Loading.vue

> I would like to submit an alternative implementation to Bitcoin Core, but I run into complications with UX because I would like Knots to be fully compatible with applications that are dependent on Core but I am afraid that this will cause conflict.

We are planning to implement "optional dependencies" for apps in a future umbrelOS update. This will allow other apps to auto-configure to whatever Bitcoin Node implementation a user has installed. But for now there is no easy way to do this other than a user manually editing docker-compose.yml files for each connecting app, or installing apps from a Community App Store that have already done this.

For now, I'd suggest that we ship this app before optional dependencies are released. We could add something to the app description that let's users know that automated integration with other apps in the Bitcoin ecosystem will be possible in the future. What do you think?

**Comment by Retropex** [comment]
Thank you for spending time on knots,

I have just [removed](https://github.com/Retropex/umbrel-bitcoin/commits/master/) all the Umbrel logos from the interface and [added](https://github.com/Retropex/umbrel-apps/commit/118f142502e155577b83d2e6896f7e0cf4071c3f) a message to warn the user that Bitcoin Knots will later be compatible with other apps.

Do I therefore have to change the Bitcoin Knots ID which is currently on the same value as Bitcoin Core (`bitcoin`) ?

Also delighted that you add the optional dependencies, it will allow the node runners to be totally sovereign on their choice of implementation of Bitcoin.

**Comment by smolgrrr** [comment]
Yep, `bitcoin-knots` to match the directory is perfect.

Also, if you've got any ideas for the 3 or so gallery images let us know :) Even, just some short slogans.

**Comment by Retropex** [comment]
Any news @smolgrrr ?

File: bitcoin-knots/exports.sh [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+export APP_BITCOIN_DATA_DIR="${EXPORTS_APP_DIR}/data/bitcoin"
+export APP_BITCOIN_RPC_PORT="8332"
+export APP_BITCOIN_P2P_PORT="8333"
+export APP_BITCOIN_TOR_PORT="8334"
+export APP_BITCOIN_ZMQ_RAWBLOCK_PORT="28332"
+export APP_BITCOIN_ZMQ_RAWTX_PORT="28333"
+export APP_BITCOIN_ZMQ_HASHBLOCK_PORT="28334"
+export APP_BITCOIN_ZMQ_SEQUENCE_PORT="28335"
+
```
Comment (smolgrrr) [inline bitcoin-knots/exports.sh:14]: Conflicts mentioned above.

**Review by smolgrrr — COMMENTED** [review]
Hi @Retropex 

Apologies for the delayed response. Thanks for adding those assets- we might redo them for you, but they're a great start.

We've just been discussing, and as highlighted above, we'll likely need to resolve the port conflict that will happen here if someone has both core and knots installed. 
We'll give you a hand with this. Like optional dependencies, we also have something in the works to resolve this too, but for now we can just assign new ports.

**Comment by Retropex** [comment]
@smolgrrr I updated the ports but I can't understand why the interface can no longer connect to the node, yet I can make RPC calls, I used this command to test:

```bash
curl --user umbrel --data-binary '{"jsonrpc": "1.0", "id": "curltest", "method": "getblockcount", "params": []}' -H 'content-type: text/plain;' http://127.0.0.1:9332/
```

**Comment by Retropex** [comment]
@smolgrrr The ports as well as the ip addresses have been updated, I did not find any conflict during my tests and I was able to install Knots and Core without conflict.

**Comment by smolgrrr** [comment]
great work mate! @nmfretz appreciate if you could give this a once over too. I'll ask the team for some help with updated assets now, but after that I think we're good to go 🚀

**Comment by nmfretz** [comment]
Awesome, thanks guys.

@Retropex are you okay if I push directly to your PR branch when reviewing/testing to get this to the finish line?

**Comment by nmfretz** [comment]
Thanks @Retropex, I have made some very small changes and this is looking good to go! We'll launch your app as soon as the updated gallery images are ready on our end (should be very soon).

------------------------
Changes:
- tweaked release notes slightly to simplify them for the initial app release on the app store
- changed some IP addresses that would end up clashing with the Lightning Node app
- changed the main UI port to 2105 so it doesn't clash with Core Lightning

File: bitcoin-knots/umbrel-app.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+dependencies: []
+repo: https://github.com/bitcoinknots/bitcoin.git
+support: https://github.com/bitcoinknots/bitcoin/issues
+port: 2105
+gallery:
+  - 1.jpg
+  - 2.jpg
+  - 3.jpg
+  - 4.jpg
+  - 5.jpg
```
Comment (nmfretz) [inline bitcoin-knots/umbrel-app.yml:41]: Note to self: change number of images once gallery is finalized

**Comment by hanb-lee** [comment]
Any update on this? It would be nice if I could get this through the default App Store. Thanks.

**Comment by yshwanth** [comment]
Push it to App store soon

**Comment by nmfretz** [comment]
Sorry for the delay due to the 1.0 launch, @Retropex. I will do a final check today and get this to the app store.

Here are the gallery images:

**Comment by nmfretz** [comment]
Thanks again for all your hard work with Bitcoin Knots @Retropex! This is now live in the official app store 🎉.

I made some final tweaks to make sure no clashes are possible with the current Bitcoin Node app when other apps go to source global environment variables.

We are currently re-doing our app ecosystem in umbrelOS now that we are on 1.0, and soon users will be able to select which Bitcoin Node app they use to integrate with other apps in the ecosystem (LND, Core Lightning, Electrs, etc).

**Comment by cyphercosmo** [comment]
Question, is it really intended for the app to start downloading the blockchain again even though I synced it when first setting up Umbrel with Bitcoin Core?

Since I have other applications relying on Bitcoin Core I can't simply uninstall it and at the same time my 1tb SSD is not enough for the blockchain x2.

**Comment by captCovalent** [comment]
29.3.knots20260508 now has BIP110 incorporated. any chance of an update?

**Comment by Retropex** [comment]
@captCovalent I will PR the update probably today or tomorrow

**Pull request merged**